### PR TITLE
[34444] Projects copy APIv3

### PR DESCRIPTION
**Pull request by oliverguenther**
- Form endpoint
  - [x] Docs
  - [X] implementation
  - [x] Basic test
  - [x] Edge cases / meta spec
  - ~~Determine default selected meta flags~~ I decided for now that the API should not magically define which modules to copy. The decision of default associations to copy is random anyway (Forums are not copied by default). In the API, all dependencies are selected for copy unless specified otherwise
  - [x] Extend schema with description using count

- Create endpoint
  - [x] Redirect to jobs status
  - [x] Extract common logic from controller
  - [x] Docs
  - [x] Basic test
  - [x] Edge cases / meta spec

- Other todos
  - [x] Take care of the send notifications param

### CODE REVIEW

File: docs/api/apiv3/endpoints/projects.apib [hunk 1]
```diff
@@ -112,8 +112,9 @@ Depending on custom fields defined for projects, additional properties might exi
 ## View project [GET]
 
 + Parameters
-    + id (required, integer, `1`) ... Project id
-
+    
++ id (required, integer, `1`) ... Project id
```
Comment (ulferts) [inline docs/api/apiv3/endpoints/projects.apib:116]: The changes in indentation will cause the parameters to not be listed again.

File: docs/api/apiv3/endpoints/projects.apib [hunk 2]
```diff
@@ -1301,6 +1304,516 @@ For more details and all possible responses see the general specification of [Fo
...
+                        "type": "Boolean",
+                        "name": "Members",
+                        "required": false,
+                        "hasDefault": true,
+                        "writable": true,
+                        "location:" "_meta"
+                    },
+                    "copyVersions": {
+                        "type": "Boolean",
+                        "name": "versions",
```
Comment (ulferts) [inline docs/api/apiv3/endpoints/projects.apib]: Typo 'versions' -> 'Versions'. 

For all the copyXYZ attributes, it would make sense to add the copy character to the name, e.g. "Copy version". This might however only be true from an API perspective, not necessarily from the perspective of our client.

File: docs/api/apiv3/endpoints/schemas.apib [hunk 3]
```diff
@@ -40,16 +40,17 @@ Only one of the links (`allowedValues`, `allowedValuesSchemas`) will exist for a
...
+| type              | The data type of the properties values                                             | MetaType     |         |
+| minLength         | The value of the property must at least contain the specified amount of characters | Integer      | 0       |
+| maxLength         | The value of the property must at most contain the specified amount of characters  | Integer      | ∞       |
+| regularExpression | The value of the property must match the given regular expression (if not null)    | String       | null    |
+| required          | If true this property is not nullable                                              | Boolean      | true    |
+| writable          | If false it is not allowed to **change** the properties value                      | Boolean      | true    |
+| location          | If present, contains a reference to the location of the property in the JSON       | String       |         |
+| description       | If present, contains a formattable, human readable description                     | Formattable  |         |
 
 All of the above properties that do not have a default value *must* be present in the schema.
```
Comment (ulferts) [inline docs/api/apiv3/endpoints/schemas.apib:55]: The addition of the `_meta` group should also be described in the `forms.apib` documenation as it will be a general pattern that clients should expect to happen in all payload objects.

File: lib/api/utilities/endpoints/bodied.rb [hunk 4]
```diff
@@ -91,6 +104,7 @@ def process(current_user, instance, params)
 
           process_service
             .new(**args.compact)
+            .with_state(process_state.call(model: instance, params: params))
```
Comment (ulferts) [inline lib/api/utilities/endpoints/bodied.rb:107]: I have been wondering for some time now whether it would make sense to make the params available within the state of a service. That way, we would not have to pass the params around and could also access them in the callbacks which could then more easily be turned into default `ActiveModel::Callbacks`.

Do you have an opinion on this, @oliverguenther ?

**Review by ulferts — CHANGES_REQUESTED** [review]

**Review by ulferts — CHANGES_REQUESTED** [review]
Sorry, pressed the wrong button so therefore the review is in two parts now. Codewise, there is nothing to complain about except some tiny things.

This is what I notice when poking around:

- [x] `"copyXYZ": true` is default for all copyable associations but setting only one `"copyXYZ": false` (or true) will set all to `false`:
image
- [x] when not providing an identifier in the body of the form request, a validation error is displayed. But the identifier property is not displayed in the payload unless a value is provided. Probably existed before.
- [x] I assume that the description on the `copyXYZ` fields is too technical to be displayed in the UI client. But we can also change this once this assumption is validated.
image
- [x] The links of the form resource do not point to the correct end points:
image
- [x] For an API client, it would be nice to also have a redirect url when the project is copied. Currently, only an html url is displayed:
image

Having that delayed/redirect flow in place in the API now looks rather promising. We have a couple of end points which can now make use of this (DELETE projects, groups, users, ...).

File: docs/api/apiv3/endpoints/projects.apib (hunk 2 above)
Comment (oliverguenther) [reply docs/api/apiv3/endpoints/projects.apib]: The current version without copy in ints label corresponds to the current labels that are used in the current project form, so I'd probably not change it.

File: lib/api/utilities/endpoints/bodied.rb (hunk 4 above)
Comment (oliverguenther) [reply lib/api/utilities/endpoints/bodied.rb:107]: Yes that sounds like a good convention.

**Review by ulferts — APPROVED** [review]
:+1: looks very good. Fine to merge once the feature spec is adapted.
